Keep the previous certificate in the bundle as well as the new one to prevent periods of service outage

arkaitzj opened the pull request:
During our tests we discovered that every time the certificates rotated we got a lot of TLS bad certificate errors in our cluster.
```2022/06/02 13:46:25 http: TLS handshake error from 10.213.162.79:60186: remote error: tls: bad certificate```

When we looked into it we discovered that the refresher sets the secret and the caBundle in the webhooks at the same time.
However, that does not account for the time it takes for a pod to get the secret updated and the operator to refresh the certificate.
During that period of time the control-plane is using the new caBundle and the operator the old one, failing all requests and provoking a service outage since none of the resources created at that time will be serviced by kubemod.

The proposed solution keeps the previous certificate in the bundle along with the new one, so that the operator can work well before and after swapping its own certificate.  
This means there are always 2 certificates in the caBundle and each certificate will be valid for 2 cron periods, if this is a concern, halving the cron period should work.

arkaitzj committed:
Add latest and current cas to the bundle

arkaitzj committed:
Support no ca

vassilvk commented:
Hi @arkaitzj,

Thank you for your pull request!

I tested it on my end and I have several comments:
- When deploying `kubemod` with your version of `kubemod-crt` on a clean cluster (no `kubemod` previously installed), the logs of `kubemod-crt` yield the cryptic message `No bundle found`. I'd rather see a log line saying `Extracting previous CA certificate...` followed by either `No previous CA certificate found`, or `[N] previous CA certificates found`
- I tested `kubemod` + your version of `kubemod-crt` with a CA cert refresh cron job running every minute.
  I found out that `kubemod` is still occasionally unhappy, issuing `remote error: tls: bad certificate` right around the time of the cert rotation. My theory is that there is still timing issue involved.
  
  The current order of renewal is:
  - cert-generate.sh
  - cert-create-secret.sh
  - cert-patch-webhooks.sh
  
  I think the issue might be resolved if we switch to:
  - cert-generate.sh
  - cert-patch-webhooks.sh
  - cert-create-secret.sh

arkaitzj committed:
Added logs

arkaitzj commented:
Hi @vassilvk,
I added the extended logging.
As for the order if I am not wrong, I don't think it matters too much because the pods can have delays of up to several minutes refreshing the configmap values, so if the refresh period is inferior to, say, 5 minutes, you will see tls errors due to the operator still using a cert from 2 periods ago. 
This could be mitigated by increasing the number of certs to keep, which is just a flag into the tool I added.

vassilvk commented:
Hi @arkaitzj, thanks, this makes sense.
Reordering without adding ugly artificial delays would likely not make a difference.
Let me test again with a 5 min cron job - if I see no `bad cert` throughout the weekend, we should be good to go with this PR.

vassilvk commented:
Closes #15
